Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -62,78 +164,188 @@ public void Initialize()
+            this.asyncLoop = this.asyncLoopFactory.Run($"Consensus Loop", async token =>
+            {
+                await this.PullerLoop(this.nodeLifetime.ApplicationStopping);
+            }, 
+            this.nodeLifetime.ApplicationStopping, 
+            repeatEvery: TimeSpans.RunOnce);
+
             this.logger.LogTrace("(-)");
         }
 
-        public IEnumerable<BlockResult> Execute(CancellationToken cancellationToken)
+        /// <summary>
+        /// Dispose components in <see cref="ConsensusLoop"/>.
+        /// </summary>
+        public void Stop()
         {
-            while (true)
+            this.asyncLoop?.Dispose();
+        }
+
+        /// <summary>
+        /// A puller method that will continuously loop and ask for the next block  in the chain from peers.
+        /// The block will then be passed to the consensus validation. 
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token that will stop the loop.</param>
+        /// <returns>The task the loop will be running under.</returns>
+        private Task PullerLoop(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
             {
-                yield return this.ExecuteNextBlock(cancellationToken);
+                try
+                {
+                    BlockItem item = new BlockItem();
+
+                    using (new StopwatchDisposable(o => this.Validator.PerformanceCounter.AddBlockFetchingTime(o)))
+                    {
+                        item.Block = this.Puller.NextBlock(cancellationToken);
+                    }
+
+                    this.logger.LogTrace("Block received from puller.");
+                    this.AcceptBlock(item);
+
+                }
+                catch (Exception ex)
+                {
+                    if (ex is OperationCanceledException)
+                    {
+                        if (this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
+                            return Task.FromException(ex);
+                    }
+
+                    // TODO Need to revisit unhandled exceptions in a way that any process can signal an exception has been
+                    // thrown so that the node and all the disposables can stop gracefully.
+                    this.logger.LogCritical(new EventId(0), ex, "Consensus loop at Tip:{0} unhandled exception {1}", this.Tip, ex.ToString());
+                    NLog.LogManager.Flush();
+                    throw;
+                }
             }
+
+            return Task.CompletedTask;
         }
 
-        public BlockResult ExecuteNextBlock(CancellationToken cancellationToken)
+        /// <summary>
+        /// A method that will accept a new block to the node.
+        /// The block will be validated and the <see cref="CoinView"/> db will be updated.
+        /// If its a new block that extends the chain then the <see cref="ConcurrentChain.Tip"/> will be set.
```

Comment:
it's

Reply:
this comment confused me, only after a while i've realized that such a block can extend the chain if we mined it (pos/pow) and it should not happen if we downloaded it from peer, i would explain this little bit more, so this is clear